clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -267,9 +267,11 @@ void Parser::ParseInnerNamespace(const InnerNamespaceInfoList &InnerNSs,
     while (!tryParseMisplacedModuleImport() && Tok.isNot(tok::r_brace) &&
            Tok.isNot(tok::eof)) {
       ParsedAttributes DeclAttrs(AttrFactory);
-      MaybeParseCXX11Attributes(DeclAttrs);
-      ParsedAttributes EmptyDeclSpecAttrs(AttrFactory);
-      ParseExternalDeclaration(DeclAttrs, EmptyDeclSpecAttrs);
+      ParsedAttributes DeclSpecAttrs(AttrFactory);
+      while (MaybeParseCXX11Attributes(DeclAttrs) ||
+              MaybeParseGNUAttributes(DeclSpecAttrs))
+        ;
```

Comment:
I’m thinking we might want to factor this out into a separate function since this exact code is now used in, like, 5 different places; maybe call it `MaybeParseCXX11AndGNUAttributes` or sth like that (just calling it `MaybeParseAttributes` would be incorrect because this doesn’t handle Microsoft attributes, so we’re probably stuck w/ the longer name unfortunately...).

Reply:
Actually, there is a TODO in ParseDeclCXX.cpp to factor out this (including the call to `ParseExternalDeclaration`) in `Parser::ParseExportDeclaration`, so now might be a good time to do that.

Reply:
If we factor out `ParseExternalDeclaration` in a function that also parses the attributes, factoring out attributes parsing is probably not as usefull right ?

Since it's quite a lot of changes, should I do it in two commits ? One fixing the bug and adding missing attribute parsing, and the other factoring everything into a `ParseExternalDeclarationWithAttrs` and replacing all uses that fit the need.

Reply:
Just pushed a new commit with what we talked about.

I'm not a big fan about the comment on the function I pushed, but I cannot find a wording that please me, I'm open to suggestion.

Reply:
Hmm, maybe just ‘Parse CXX11 and GNU attributes followed by a declaration’.

Reply:
Yes, but it could be understood the wrong way around I think...
Let's wait on feedback from others, maybe someone will have an eureka moment.